### PR TITLE
SHARD-2024: Fix incorrect transaction index

### PR DESCRIPTION
**Pull request by jintukumardas**
### **User description**
Correct the incorrect transactionIndex output for `eth_getTransactionByBlockHashAndIndex` and `eth_getTransactionByBlockNumberAndIndex`.

___

### **PR Type**
Bug fix

___

### **Description**
- Fix incorrect transaction index in API responses.

- Add undefined check for transaction index.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>api.ts</strong><dd><code>Correct transaction index and add undefined check</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/api.ts

<li>Correct transaction index calculation in API methods.<br> <li> Add check for undefined transaction index.

</details>

  </td>
  <td><a href="https://github.com/shardeum/json-rpc-server/pull/139/files#diff-769911c416ccf8514d8fd941ae0abe8fb5c606ade0c218e22151a5f5f9f3d700">+6/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about PR-Agent usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

**Comment by shanejonas** [comment]
Can we investigate why it has `transactionIndex: "0x1"` from where its getting this from? this fix just overriding it without any explanation on the root cause. If its incorrect from the json-rpc-server ingestion or the collector API (or some other downstream service) we should find/fix it there.

**Review by urnotsam — DISMISSED** [review]

**Comment by urnotsam** [comment]
@jintukumardas @shanejonas 

its hard coded in like 8 places in shardeum server.  Not sure what the downstream changes frmo that point on would be

**Comment by BelfordZ** [comment]
Did anyone actually check to see if its returning proper TX indexes now?

I can't see how this would fix the problem.

**Comment by jintukumardas** [comment]
Thanks for reviewing. This solution aligns with our existing approach for explorer responses where we override the value with user given index (see: [https://github.com/shardeum/json-rpc-server/blob/26a7124583f4406ce7a95fc63d1bca8b76acac95/src/api.ts#L221]). As Sam noted, this value is hardcoded everywhere and represents the transaction's index within the requested block. This allows for a quick fix since we are already fetching that index from the block response. I've locally verified these changes by pointing RPC to stagenet. The attached image shows a comparison: the left side displays block data from the explorer, and the right side shows the 3rd indexed transaction for the same block hash, confirming they match.

Screenshot 2025-03-20 at 10 16 54 PM

**Comment by BelfordZ** [comment]
# ETH Handlers Testing

## Summary of work done
- Created comprehensive test suites for two JSON-RPC handlers:
  - `eth_getTransactionByBlockHashAndIndex`
  - `eth_getTransactionByBlockNumberAndIndex`
- Fixed TypeScript errors in the handlers
- Identified and documented issues with callback usage and return values
- Achieved > 93% code coverage for both handlers

## Test Coverage
- **eth_getTransactionByBlockHashAndIndex**: 93.33% statements, 80.76% branches, 100% functions, 97.05% lines
- **eth_getTransactionByBlockNumberAndIndex**: 93.58% statements, 82.14% branches, 100% functions, 97.10% lines

## Issues Found
Two main issues were identified and documented as tickets:

1. **Multiple callback calls** - Both handlers call the callback function multiple times in certain code paths, which can lead to unexpected behavior for clients.

2. **Inconsistent return values** - When transactions aren't found, handlers sometimes return `undefined` instead of the expected `null` value specified in the Ethereum JSON-RPC specification.

## Test Scenarios
The tests cover all major code paths and edge cases:
- Input validation failures
- Successful transaction retrieval via Collector API
- Transaction index adjustment
- Error handling for Collector API exceptions
- Explorer API fallback behavior
- Special block number values ('latest', 'earliest')
- Error handling for Explorer API
- Configuration state handling (disabled components)

## Next Steps
- Address the identified issues in the tickets folder
- Implement similar test patterns for other ETH handlers
- Consider refactoring handlers to reduce code duplication and improve maintainability

**Review by urnotsam — DISMISSED** [review]

**Review by mhanson-github — APPROVED** [review]

**Review by mssabr01 — COMMENTED** [review]
Please remove the anys from the function prototypes in BuildGetTransactionByBlockHashAndIndex and BuildGetTransactionByBlockNumberAndIndex

File: src/eth-handlers/eth_getTransactionByBlockHashAndIndex.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+      .update(api_name + Math.random() + Date.now())
+      .digest('hex')
+    logEventEmitter.emit('fn_start', ticket, api_name, performance.now())
+    /* prettier-ignore */ if (firstLineLogs) { console.log('Running eth_getTransactionByBlockHashAndIndex', args) }
+
+    logEventEmitter.emit('fn_end', ticket, { success: true }, performance.now())
+    let result: string | completeReadableReceipt | null | undefined = null
+
+    try {
+      const blockResp = await collectorAPI.getBlock((args as any)[0], 'hash', true)
```
Comment (mssabr01) [inline src/eth-handlers/eth_getTransactionByBlockHashAndIndex.ts:57]: Can we make this work without any?

Comment (mssabr01) [reply src/eth-handlers/eth_getTransactionByBlockHashAndIndex.ts:57]: Same question for all the `as any` below
